### PR TITLE
[SPARK-42531][CONNECT] Scala Client Add Collections Functions

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR adds all the collections functions to `functions.scala` for Scala client. This is the last PR large functions PR, there are a few functions missing, these will be added later.

### Why are the changes needed?
We want the Scala client to have API parity with the existing API

### Does this PR introduce _any_ user-facing change?
Yes, it adds functions to the Spark Connect Scala Client.

### How was this patch tested?
Added tests to `PlanGenerationTestSuite` and to `ProtoToPlanTestSuite`. I have added a few tests to `ClientE2ETestSuite` for lambda functions (to make sure name scoping works) and the array shuffle function (non-deterministic, hard to test with golden files).

### CODE REVIEW

**Comment by hvanhovell** [comment]
For the reviewer I have manually tested this with scala-2.13.

**Comment by hvanhovell** [comment]
cc @LuciferYang

**Comment by LuciferYang** [comment]
Checked the new test with Scala 2.13, all passed

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala [hunk 1]
```diff
@@ -336,4 +336,31 @@ class ClientE2ETestSuite extends RemoteSparkSession {
     checkSample(datasets.get(2), 3.0 / 10.0, 6.0 / 10.0, 9L)
     checkSample(datasets.get(3), 6.0 / 10.0, 1.0, 9L)
   }
+
+  test("lambda functions") {
+    // This test is mostly to validate lambda variables are properly resolved.
+    import org.apache.spark.sql.functions._
```
[the remark this answers is not in the document]
Comment (hvanhovell) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/ClientE2ETestSuite.scala:342]: I guess we can.

**Review by LuciferYang — APPROVED** [review]
+1, LGTM (pending CI)

**Pull request closed without merging**